### ISSUE TITLE
AzureFunctionAppV1 fails due to missing permission to perform joinViaServiceEndpoint/action on subnet

### ISSUE THREAD

**Issue by rantsi** [microsoft/azure-pipelines-tasks#16167]
## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: AzureFunctionAppV1

list here (V# not needed):  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

## Environment

Private agent Agent.Version: 2.200.2 Agent.OSVersion: 10.0.17763

## Issue Description

During the deployment the build prints a warning and fails:
`##[warning]Failed to update App Service configuration details. Error: Error: Failed to update App service 'func-xxx-dev' configuration. Error: The client 'xxxx...' with object id 'xxxx...' has permission to perform action 'Microsoft.Web/sites/config/write' on scope '/subscriptions/xxx.../resourceGroups/rg-xxx/providers/Microsoft.Web/sites/func-xxxx-dev/config/web'; however, it does not have permission to perform action 'joinViaServiceEndpoint/action' on the linked scope(s) '/subscriptions/xxx.../resourceGroups/rg_xxx/providers/Microsoft.Network/virtualNetworks/vnet-xxx-dev/subnets/snet-xxx-dev' or the linked scope(s) are invalid. (CODE: 403)` 

yaml for step:
```
- task: AzureFunctionApp@1 
  displayName: 'Deploy Azure Function App' 
  inputs: 
    azureSubscription: '$(Parameters.AzureSubscription)' 
    appType: '$(Parameters.AppType)' 
    appName: '$(Parameters.AppName)'
```

Using the same credentials to publish the same package with [PowerShell](https://docs.microsoft.com/en-us/azure/azure-functions/deployment-zip-push#with-powershell) using _Publish-AZWebApp_ completes successfully.

**Comment by FinVamp1** [comment]
Hi, @rantsi  do you have more information about this scenario? Are you publishing to a Windows App or a Linux one? Is the Storage account used for this Azure Function App on a subnet with the Azure Function App that has Service Endpoints enabled?

**Comment by annieEllenberger** [comment]
We are also encountering this issue with our Function Apps that have inbound "Access Restrictions" configured. 

A few more details:
- Hosting on Windows (Azure Functions v4; .NET 6.0; dotnet-isolated)
- Networking rules on the Function App look like this:
![image](https://user-images.githubusercontent.com/62848777/196788961-0332f00f-c9cb-4e76-b15c-4d0ceb6309e2.png)
- The corresponding Storage Account for the Function App has network rules configured as the following:
![image](https://user-images.githubusercontent.com/62848777/196802185-d9edec9b-3420-4009-a418-85401c0ec47c.png)

The following is displayed in the logs of the deploy task in Azure DevOps pipelines: 
```##[warning]Failed to update App Service configuration details. Error: Error: Failed to update App service '<FunctionAppName>' configuration. Error: The client '<GUID>' with object id '<GUID>' has permission to perform action 'Microsoft.Web/sites/config/write' on scope '/subscriptions/<GUID>/resourceGroups/<RGName>/providers/Microsoft.Web/sites/<FunctionAppName>/config/web'; however, it does not have permission to perform action 'joinViaServiceEndpoint/action' on the linked scope(s) '/subscriptions/<GUID>/resourceGroups/<RGName>/providers/Microsoft.Network/virtualNetworks/<VNetName>/subnets/<subnetName>' or the linked scope(s) are invalid. (CODE: 403)```

... but it doesn't actually *fail* the pipeline (says it's _both_ a Warning AND an Error?)

The code package appears to be deployed correctly. The application is completely functional. App Settings ("Configuration Settings") on the Function App are successfully created/modified by the pipeline. 

So, is there actually an issue occurring behind the scenes? Or is this a misleading non-issue within the `AzureFunctionApp@1` task? 

**Note**: we have a second Function App (same specifications as the one above) that is deployed using the same exact pipeline. The only difference between the two Function Apps, is that the second does *not* have any _inbound_ "Access Restrictions" (only outbound VNet integration). Both Function Apps share the same storage account. The warning/error does NOT appear for the second Function App (the one with only outbound network rules).

**Comment by annieEllenberger** [comment]
Also seeing the following error while attempting to view the log stream of BOTH function apps mentioned in my previous comment (even though functionality appears to be 'working' at the same time): 
![image](https://user-images.githubusercontent.com/62848777/196977785-e6ad6742-091f-45ad-b9e8-0fc7191e52fb.png)

**Comment by FinVamp1** [comment]
Hi, assigned this to me to look at.

**Comment by jochemdevos** [comment]
We also see this issue when deploying function app, no 'functional' impact except for the error message and failed job state. Any findings?

**Comment by FinVamp1** [comment]
Hi, I'm still trying to reproduce this. Will update this by the end of this week.

**Comment by jochemdevos** [comment]
This must really be at the bottom of the pile ;)

**Comment by FinVamp1** [comment]
Hello, this is something I will continue to try and get to the bottom of. I think this issue is similar to https://github.com/microsoft/azure-pipelines-tasks/issues/15961 and revolves around the permissions the Service Principal from the Service Connection has on the underlying Web App. If it has Contributor or great it should able to perform the joinViaServiceEndpoint/action. I've been trying to repro and will spend more time tomorrow manually registering a Service Principal with reduced pemissions. Can you confirm what what permissions your Service Principal has?

**Comment by annieEllenberger** [comment]
@FinVamp1  In our case, the Service Principal from the Service Connection is a "Contributor" on the Resource Group that contains the Function Apps and their respective Storage Accounts. The VNet with the subnets utilized by the Function Apps are in a completely separate Resource Group, to which the Service Principal does _not_ appear to have any role assignment. So, perhaps this is the reason for the warning/error message? If so, I am still wondering:
1. Why this occurs even though there does not appear to be any real/functional/tangible impact? 
2. What is the appropriate way to grant the Service Principal access, in a secure fashion? Role assignments can be made at the RG or VNET resource level... but granting access to the entire VNnet would be far more access than should be necessary for this use case (and our Security Team would most likely have issues with this)

**Comment by erkkiholttinen** [comment]
We are encountering this same problem with an Azure Managed Application. In our situation the app services and function apps are located in the managed resource group in which we have owner rights. The virtual network is in the other hand managed by the client in a separate resource group which to we do not have any access.

This now occurs in a situation where we are applying inbound traffic rules to existing applications. Have not yet tried if this would work in case those rules would be applied with the initial deployment when the customer buys the Managed Application.

**Comment by FinVamp1** [comment]
Hello, I've checked with our VNet team. 

1. This issue happens when there's an ARM PUT\PATCH on the site which has access restrictions defined. This happens as the Azure Functions task will do a PUT for the App Settings on the Site. As it's not changing the access restrictions it's just an error back from the ARM call.
2. The Service Principal could have the join permissions on the subnets mentioned in the access restriction rules, but usually these roles are separated in large orgs as you have pointed. 

Let me know if calling this out in the task logs would help diagnose these issues?

**Comment by FinVamp1** [comment]
Closing this as there's no action on the Task owners that could help here unless we add a warning. 

This issue happens when there's an ARM PUT\PATCH on the site which has access restrictions defined. This happens as the Azure Functions task will do a PUT for the App Settings on the Site. As it's not changing the access restrictions it's just an error back from the ARM call.
The Service Principal could have the join permissions on the subnets mentioned in the access restriction rules, but usually these roles are separated in large orgs as you have pointed.

**Comment by krysickij1** [comment]
@FinVamp1 
I faced the same issue with the AzureFunctionApp@2.

I was trying to deeploy the code to Windows functionapp with powershell (it was not working even with the empty project). Eventually I was able to reproduce the error on other function app when after I added vnet/subnet to the access restriction. The problem disappeared when I removed them.  I don't know why but somehow this task is trying to change the subnets mentioned in the access restriction. With service principal with privileges on the subscription level it was not causing any errors, but when I was using resource-group level one it actually returned the error.
